File: Atividade_5/Animal.py
class Animal:
    def __init__(self, especie, nome):
        self.especie = especie
        self.nome = nome

    def emitir_som(self):
        return ""

    def informacoes(self):
        print(f"Nome: {self.nome}, Espécie: {self.especie}")


class Cachorro(Animal):
    def __init__(self, nome):
        super().__init__("Cachorro", nome)

    def emitir_som(self):
        return "Au Au"


class Gato(Animal):
    def __init__(self, nome):
        super().__init__("Gato", nome)

    def emitir_som(self):
        return "Miau"


def mostrar_sons(animais):
    for animal in animais:
        animal.informacoes()
        print(f"{animal.nome} diz: {animal.emitir_som()}")


lista_de_animais = [Cachorro("Rex"), Gato("Mimi")]
mostrar_sons(lista_de_animais)

File: Atividade_5/Animal.go
package main

import "fmt"

type Animal interface {
	emitirSom() string
	informacoes()
}

type animal struct {
	especie string
	nome    string
}

func (a *animal) informacoes() {
	fmt.Printf("Nome: %s, Espécie: %s\n", a.nome, a.especie)
}

type Cachorro struct {
	animal
}

func newCachorro(nome string) *Cachorro {
	return &Cachorro{animal{especie: "Cachorro", nome: nome}}
}

func (c *Cachorro) emitirSom() string {
	return "Au Au"
}

type Gato struct {
	animal
}

func newGato(nome string) *Gato {
	return &Gato{animal{especie: "Gato", nome: nome}}
}

func (g *Gato) emitirSom() string {
	return "Miau"
}

func mostrarSons(animais []Animal) {
	for _, animal := range animais {
		animal.informacoes()
		fmt.Printf("%s diz: %s\n", animal.(*animal).nome, animal.emitirSom())
	}
}

func main() {
	animais := []Animal{
		newCachorro("Rex"),
		newGato("Mimi"),
	}
	mostrarSons(animais)
}
